doc/appendices/development.rst

```diff
@@ -8,4 +8,400 @@
+
+Data management must be isolated from data use for any code that is to run on
+the device. This
+allows low-level physics classes to operate on references to data using the
+exact same device/host code. Furthermore, state data (one per track) and
+shared data (definitions, persistent data, model data) should be separately
+allocated and managed.
+
+Store
+  Generic name for a class that manages GPU data by means of a host class,
+  using ``celeritas::DeviceVector`` (or ``thrust`` or ``VecGeom`` wrappers as
+  needed) to manage the on-device data. Use DeviceVectors for containers that
+  don't need special initialization (i.e. have "plain old data").
+
+Params (model parameters)
+  Provide a CPU-based interface to manage and provide access to constant shared
+  GPU data, usually model parameters or the like. The Params class itself can
+  only be accessed via host code, but it should use the
+  ``celeritas::DeviceVector`` or ``thrust`` or ``VecGeom`` wrappers to manage
+  on-device data. A params class can contain metadata (string
+  names, etc.) suitable for host-side debug output and for helping related
+  classes convert from user-friendly input (e.g. particle name) to
+  device-friendly IDs (e.g. particle def ID).
+
+State (state variables)
+  Thread-local data specifying the state of a single particle track with
+  respect to a corresponding model (``FooParams``). The state
+  data resides on device but is managed by a host class ``FooStateStore`` using
+  ``DeviceVector`` or the like. It is an implementation detail whether the
+  state data is stored as a struct of arrays (SOA) or an array of structs
+  (AOS), but if stored as AOS then the per-track state struct should be named
+  ``TrackFooState``.
+
+Data
+  A standalone, plain-old-data struct to data owned by another class (e.g., a
+  Params class) but stored on the GPU. This struct is used to transfer GPU
+  pointers and other kernel parameters between host and device. A Data
+  struct can hold other Data structs as data members. Inside unit tests for
+  debugging, Data can reference *host* data if the corresponding functions
+  being called are also on-host. Defining Data structs in separate files
+  from the memory management classes means that NVCC doesn't have to include
+  those headers, speeding up compilation and perhaps allowing the host code to
+  use newer implementations of the C++ standard.
+
+TrackView
+  Device-instantiated class that provides read/write access to the data for a
+  single track, in the spirit of `std::string_view` which adds functionality to
+  data owned by someone else. It combines the state variables and model
+  parameters into a single class. The constructor always takes const references
+  to ParamsData and StatePointers as well as the track ID. It encapsulates
+  the storage/layout of the state and parameters, as well as what (if any) data
+  is cached in the state.
+
+View
+  Device-instantiated class with read/write access for data shared across
```

Comment:
Is the 'Device' limitation still in place?  I.e. is code using Views not able to be tested on CPU?

Reply:
Nope this was just a missed update from the year+-old document :) fixed